### ISSUE TITLE
speed up bookdown/rmarkdown/knitr by parallel computing (probably just parallel compilation)

### ISSUE THREAD

**Issue by maxheld83** [rstudio/bookdown#307]
I love how much faster `bookdown::gitbook()` is relative to LaTeX compilation, but it can still be quite slow, even with caching etc.
I'm on a multicore machine, but bookdown (and the rest of the toolchain) only appears to use one core.

So I was wondering whether parts of this could be paralelized.

Of course, parallelising is also an issue for just what people write *inside* the chunks, and that's not `bookdown`'s problem.
Also, I guess that computing several chunks *in parallel* is near impossible because of possibly mutual dependencies, and detecting these (though the `autodep` option does something like this, right?).

One thing I wondered might work was just to compile the different *output formats* in parallel (say, LaTeX and gitbook). Would that be feasible?

Other ideas of what could be parallelised?

**Comment by yihui** [comment]
There is certainly room for parallelization, but I (or you) need to profile the compilation and figure out the real slow parts before I can decide whether it makes sense to go parallel.

When you work on a book, you can preview a single chapter, which is the default behavior when you click the Knit button: https://bookdown.org/yihui/bookdown/rstudio-ide.html I think previewing a chapter should be fast enough in most cases.

**Comment by gorkang** [comment]
Maybe using one core for each of the rmd_files in the _bookdown.yml could be feasible?

If possible, that would speed up the rendering quite a bit.

**Comment by mathematiguy** [comment]
Mainly I'd like something like, storing chapters as `.Rmd` files that can be rendered independently into a folder, with maybe one more step to build the whole book from the already rendered files (if necessary). In this case, it would be easy enough to parallelise the rendering of all files using `make`, and I could manage the cpu + ram use of each chapter myself.

But, this is still an excellent project and I plan to use it plenty. I just hope that it'll be possible to keep it up to date more easily. Some of the chapters I plan to produce will take minutes to render because of model complexity, and I have the cores available to speed this up significantly.

**Comment by bakaburg1** [comment]
Parallelization could be helpful also for one document .Rmd file processing, by running the chunks in parallel.
I often create very large reports with hundreds of chunks that create images (I generate first an intermediary file with spin()), and it takes a lot to render.
Chunks could be run in parallel, saved, and then stitched together with the text part of the document.

**Comment by mathematiguy** [comment]
I haven’t tried it yet, but Drake (https://github.com/ropensci/drake) can be used to do that.

I want to try it out pretty badly, but I haven’t settled on a structure that will work for me just yet.

**Comment by yihui** [comment]
@bakaburg1 That requires the assumption that the code chunks are independent (i.e. "embarrassingly parallel"). This assumption is not generally true, but I agree when it is true, it could significantly speed up the knitting process.

**Comment by bakaburg1** [comment]
Yep I thought about as soon I ended writing the post. But btw is the same
assumption one makes when using mclapply or other parallel solutions. It
could be a chunk option to be explicitly declared, and if it is designed
and explained clearly it could be helpful in many settings.

Il giorno ven 24 mag 2019 alle 18:24 Yihui Xie <notifications@github.com>
ha scritto:

**Comment by wikiselev** [comment]
Is this going to be considered for development at all? We are facing the same issue as described by @bakaburg1. We have a huge Bioinformatics course (https://github.com/hemberg-lab/scRNA.seq.course) and at the moment it takes ~2 hours to compile (it does quite a lot of calculations). Because of this we had to move away from Travis and build it manually on premise. Parallelisation would allow us to finish in 10-20 mins. The course is still growing and I expect the compiling time will go up to 3-4 hours in the future...

**Comment by mathematiguy** [comment]
It wouldn’t be as pleasant, but you could refactor your notebooks to use drake for build management, which would get you that 20 min build time. 

But it might take a long time to do the rewrite, it kinda depends.

**Comment by maxheld83** [comment]
Just want to support what @mathematiguy said here a couple of times: if the computation *inside your* chunks is expensive, the more powerful [drake](https://github.com/ropensci/drake) pkg might be your ticket.

My original suggestion here (since I was the OP :) was to parallelise *different output formats* (which are currently rendered in sequence), which is:
- arguably a much simpler / perhaps almost embarrassingly parallel problem?
- particular to bookdown/rmarkdown, *not* the computation *inside* the chunks.

Parallelising the output formats is also unlikely to give much of a performance boost to substantively expensive reports (if any), so this could be a lot of work for relatively little gain, which I think is why @yihui rightly closed this.

So bottom line: if your project has expensive chunks, we might not want to look for a solution in the reporting tools. 

Separation of concerns would suggest to look at other pkgs/approaches for computationally expensive R, such as:

- drake
- parallelisation
- [cloud computing](https://cloudyr.github.io)

**Comment by yihui** [comment]
@wikiselev Have you tried to cache the time-consuming code chunks? https://bookdown.org/yihui/bookdown/some-tips.html

**Comment by wikiselev** [comment]
Thanks for the replies, everyone! Unfortunately, it does not look like either drake or bookdown caching will work for us. This is because the compilation is always happening in a new docker container which is destroyed after the process is finished. (maybe I am wrong here...)

Anyway, it looks like a complicated problem and I appreciate your feedback and understand that it requires lots of time to dive into this.

**Comment by yihui** [comment]
@wikiselev You can save the cache folder somewhere after the book has been compiled (e.g. upload to an S3 bucket or push to a Git repo), and reuse it the next time.

**Comment by wikiselev** [comment]
That's very useful to know, thanks @yihui ! I will try and will let you know.

**Comment by wikiselev** [comment]
@yihui one more question regarding the `_bookdown_files`. It looks like it contains all of the generated plots. So what is saved in the `cache` folder then? And shall I upload both `_bookdown_files` and `cache` to S3?

**Comment by yihui** [comment]
@wikiselev The `_bookdown_files` folder contains both `*_cache/` and `*_files/`. I suggest you just upload the whole `_bookdown_files` folder, although the plots under `*_files/` not generated from cached code chunks could be ignored.
